### PR TITLE
Design and implement message field for tx table

### LINKED ISSUE #931: Design and implement message field for tx table
This is a JSON-like arbitrary field that will be different depending on the message. It should have enough info to for frontend to render the transaction and message information. Should handle
- MsgSend
- MsgCreateDataSource
- MsgEditDataSource
- MsgCreateOracleScript
- MsgEditOracleScript
- MsgRequest
- MsgReport
- MsgAddOracleAddress
- MsgRemoveOracleAddress

### PR DESCRIPTION
**Pull request by Benzbeeb**
Fixed #931

### CODE REVIEW

File: chain/db/db.go [hunk 1]
```diff
@@ -107,23 +111,136 @@ func (b *BandDB) HandleTransaction(tx auth.StdTx, txHash []byte, logs sdk.ABCIMe
...
+		return fmt.Sprint(string(jsonString)), nil
 	case zoracle.MsgEditDataSource:
-		return b.handleMsgEditDataSource(txHash, msg, events)
+
+		err = b.handleMsgEditDataSource(txHash, msg, events)
+		if err != nil {
+			return "", err
+		}
+
+		jsonMap["dataSourceId"] = events["edit_data_source.id"]
```
Comment (sorawit) [inline chain/db/db.go]: Don't type out event name directly. Use the names defined in `zoracle` module.

File: chain/db/db.go [hunk 2]
```diff
@@ -107,23 +111,136 @@ func (b *BandDB) HandleTransaction(tx auth.StdTx, txHash []byte, logs sdk.ABCIMe
...
+		jsonMap["type"] = events["message.action"]
+		jsonString, err := json.Marshal(jsonMap)
+		if err != nil {
+			return "", err
+		}
+
+		return fmt.Sprint(string(jsonString)), nil
```
Comment (sorawit) [inline chain/db/db.go]: Let's return this after the switch? This is crazy repetition.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/app/db_app.go [hunk 3]
```diff
@@ -130,6 +130,7 @@ func (app *dbBandApp) DeliverTx(req abci.RequestDeliverTx) (res abci.ResponseDel
 	if stdTx, ok := tx.(auth.StdTx); ok {
 		txHash := tmhash.Sum(req.Tx)
 
+		messages := app.dbBand.HandleTransaction(stdTx, txHash, logs)
```
Comment (taobun) [inline chain/app/db_app.go]: Please make it into 2 step
1. Create transaction without message.
Handle transaction
2. Update msgs to transaction later

File: chain/db/db.go [hunk 4]
```diff
@@ -107,23 +111,76 @@ func (b *BandDB) HandleTransaction(tx auth.StdTx, txHash []byte, logs sdk.ABCIMe
...
+
+	case zoracle.MsgCreateOracleScript:
+
+		jsonMap["oracleScriptID"] = events[zoracle.EventTypeCreateOracleScript+"."+zoracle.AttributeKeyID]
+
+	case zoracle.MsgEditOracleScript:
+
+	case zoracle.MsgRequestData:
+
+		jsonMap["requestID"] = events[zoracle.EventTypeRequest+"."+zoracle.AttributeKeyID]
```
Comment (taobun) [inline chain/db/db.go:306]: We need oraclescriptName

File: chain/db/db.go [hunk 5]
```diff
@@ -107,23 +111,76 @@ func (b *BandDB) HandleTransaction(tx auth.StdTx, txHash []byte, logs sdk.ABCIMe
...
+
+	case zoracle.MsgReportData:
+
+	case zoracle.MsgAddOracleAddress:
```
Comment (taobun) [inline chain/db/db.go:312]: ValidatorMoniker

File: chain/db/db.go [hunk 6]
```diff
@@ -107,23 +111,76 @@ func (b *BandDB) HandleTransaction(tx auth.StdTx, txHash []byte, logs sdk.ABCIMe
...
+
+	case zoracle.MsgRemoveOracleAdderess:
```
Comment (taobun) [inline chain/db/db.go]: ValidatorMoniker

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/app/db_app.go [hunk 7]
```diff
@@ -156,9 +157,15 @@ func (app *dbBandApp) DeliverTx(req abci.RequestDeliverTx) (res abci.ResponseDel
 			stdTx.GetSigners()[0],
 			true,
 			app.DeliverContext.BlockHeight(),
+			"[]",
+		)
+
+		messages := app.dbBand.HandleTransaction(stdTx, txHash, logs)
+		app.dbBand.UpdateTransaction(
```
Comment (taobun) [inline chain/app/db_app.go]: Why this code doesn't be in HandleTransaction?

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/db/db.go [hunk 8]
```diff
@@ -220,31 +226,88 @@ func (b *BandDB) HandleTransaction(tx auth.StdTx, txHash []byte, logs sdk.ABCIMe
...
+		jsonMap["oracleScriptName"] = oracleScript.Name
+		jsonMap["requestID"] = events[zoracle.EventTypeRequest+"."+zoracle.AttributeKeyID]
+
+	case zoracle.MsgReportData:
+
+	case zoracle.MsgAddOracleAddress:
+		val, _ := b.StakingKeeper.GetValidator(b.ctx, msg.Validator)
+		jsonMap["validatorMoniker"] = val.Description.Moniker
+
+	case zoracle.MsgRemoveOracleAdderess:
```
Comment (taobun) [inline chain/db/db.go]: Rebase master?

**Review by taobun — CHANGES_REQUESTED** [review]

**Review by taobun — DISMISSED** [review]

File: chain/app/db_app.go [hunk 9]
```diff
@@ -156,9 +157,11 @@ func (app *dbBandApp) DeliverTx(req abci.RequestDeliverTx) (res abci.ResponseDel
 			stdTx.GetSigners()[0],
 			true,
 			app.DeliverContext.BlockHeight(),
+			make([]map[string]interface{}, 0),
```
Comment (sorawit) [inline chain/app/db_app.go]: Can we put nil?

File: chain/db/db.go [hunk 10]
```diff
@@ -220,31 +226,91 @@ func (b *BandDB) HandleTransaction(tx auth.StdTx, txHash []byte, logs sdk.ABCIMe
...
+		newMsg, err := b.HandleMessage(txHash, msg, kvMap)
 		if err != nil {
 			panic(err)
 		}
+
+		messages = append(messages, newMsg)
+
 	}
+
+	b.UpdateTransaction(
```
Comment (sorawit) [inline chain/db/db.go]: nit: I think this can be one line.

File: chain/db/db.go [hunk 11]
```diff
@@ -220,31 +226,91 @@ func (b *BandDB) HandleTransaction(tx auth.StdTx, txHash []byte, logs sdk.ABCIMe
...
+		return nil, err
+	}
+
+	err = json.Unmarshal(rawBytes, &jsonMap)
+	if err != nil {
+		return nil, err
+	}
+
 	switch msg := msg.(type) {
 	// Just proof of concept
```
Comment (sorawit) [inline chain/db/db.go]: Let's remove this comment.

File: chain/db/transaction.go [hunk 12]
```diff
@@ -37,7 +40,12 @@ func (b *BandDB) AddTransaction(
 	sender sdk.AccAddress,
 	success bool,
 	blockHeight int64,
+	messages []map[string]interface{},
```
Comment (sorawit) [inline chain/db/transaction.go]: Why do we take this parameter if we are always passing nil?

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/db/table.go [hunk 13]
```diff
@@ -66,14 +68,15 @@ type OracleScriptRevision struct {
...
-	BlockHeight int64  `gorm:"not null"`
+	TxHash      []byte          `gorm:"primary_key"`
+	Timestamp   int64           `gorm:"not null"`
+	GasUsed     int64           `gorm:"not null"`
+	GasLimit    uint64          `gorm:"not null"`
+	GasFee      string          `gorm:"not null"`
+	Sender      string          `gorm:"not null"`
+	Success     bool            `gorm:"not null"`
+	BlockHeight int64           `gorm:"not null"`
+	Messages    json.RawMessage `sql:"json"`
```
Comment (sorawit) [inline chain/db/table.go]: Does it need to be not null?

**Review by sorawit — CHANGES_REQUESTED** [review]

**Review by taobun — APPROVED** [review]

**Review by sorawit — APPROVED** [review]

**Pull request merged**